### ISSUE TITLE
Help with decorating a http call with circuitbreaker, threadpoolbiulkhead and timelimiter

### ISSUE THREAD

**Issue by scott9999** [resilience4j/resilience4j#1460]
Resilience4j version: 1.7.0

Java version: 11

Thanks for raising a Resilience4j issue.

Hi, I am attempting to replace Hystrix with Resilience4j. So far I have not been able to decorate a method that makes a HTTP call with a circuitbreaker, threadpoolbulkhead and timelimiter, and have the thread interrupted when the timelimiter times out. Is this possible, if so can you please provide some guidance.

Many thanks

**Comment by hexmind** [comment]
Yes, sure. There it is https://resilience4j.readme.io

**Comment by scott9999** [comment]
Hi,

Thanks for the quick response. I could not see from there how to configure the timelimiter to interrupt the thread that is running. The Timelimiter times the request out but does not interrupt the thread that is making the http call.


________________________________

**Comment by RobWin** [comment]
I assume you are using a CompletableFuture? CompletableFuture cannot be cancelled

**Comment by scott9999** [comment]
I am using a CompletableFuture. Would a Future be cancellable ?
________________________________

**Comment by RobWin** [comment]
Yes, future is cancellable

**Comment by scott9999** [comment]
Hi,

I have not been able to find a way to get the future cancelled (the thread interrupted), is there an example of this. I have CircuitBreaker, ThreadPoolBulkHead and TimeLimiter.

Regards,
Scott
________________________________

**Comment by RobWin** [comment]
Could you please show me your code snippet and configuration?

**Comment by scott9999** [comment]
Hi,

Below is a copy of the code I am testing:

public  TestResilienceInterrupt() {
    ThreadPoolBulkheadConfig threadPoolBulkheadConfig = ThreadPoolBulkheadConfig.custom().coreThreadPoolSize(2).maxThreadPoolSize(2).build();
    threadPoolBulkhead = ThreadPoolBulkhead.of("threadPoolBulkhead", threadPoolBulkheadConfig);

    TimeLimiterConfig timeLimiterConfig = TimeLimiterConfig.custom().timeoutDuration(Duration.ofSeconds(10)).cancelRunningFuture(true).build();
    timeLimiter = TimeLimiter.of("timelimiter", timeLimiterConfig);

    CircuitBreakerConfig circuitBreakerConfig = CircuitBreakerConfig.ofDefaults();
    circuitBreaker = CircuitBreaker.of("circuitbreaker", circuitBreakerConfig);
    scheduledExecutorService = Executors.newScheduledThreadPool(3);
}

public void testBasic() throws Exception {

    Supplier<Future<String>> supplier = () -> waitForInterrupt();

    CompletableFuture<Future<String>> future = Decorators.ofSupplier(supplier)
            .withThreadPoolBulkhead(threadPoolBulkhead)
            .withTimeLimiter(timeLimiter, scheduledExecutorService)
            .withCircuitBreaker(circuitBreaker)
            .get().toCompletableFuture();
    logger.info(future.get().get());
}

public Future<String> waitForInterrupt() {
    return new FutureTask(() -> {
        logger.info("running waitForInterrupt");
        try {
            logger.info("sleeping waitForInterrupt");
            Thread.sleep(30000);
        } catch (Exception e) {
            logger.info("sleeping waitForInterrupt was interrupted");
        }
        logger.info("finished waitForInterrupt");
        return "finished";
    });
}

The waitForInterrupt is not called. This copy of waitForInterrupt is called but is not interrupted:


public Future<String> waitForInterrupt() {
    logger.info("running waitForInterrupt");
    try {
        logger.info("sleeping waitForInterrupt");
        Thread.sleep(30000);
    } catch (Exception e) {
        logger.info("sleeping waitForInterrupt was interrupted");
    }
    logger.info("finished waitForInterrupt");
    return CompletableFuture.completedFuture("Finished");
}



________________________________

**Comment by RobWin** [comment]
Yes, as I said.
ThreadPoolBulkhead always creates a CompletableFuture and CompletableFuture is not cancelable.

**Comment by scott9999** [comment]
Hi Robert,

Thank you for the example code, I have tested and the timeout interrupts the thread.

But I think I am misunderstanding something, I have tried a few examples with the timelimiter, for example:

private <T> CompletableFuture<T> executeAsync(Supplier<T> supplier){
        return Decorators.ofSupplier(supplier)
                .withThreadPoolBulkhead(threadPoolBulkhead)
                .withTimeLimiter(timeLimiter, scheduledExecutorService)
                .withCircuitBreaker(circuitBreaker)
                .withRetry(retry,scheduledExecutorService)
                .get().toCompletableFuture();
    }

and the spring cloud circuit breaker with a timelimiter. If a supplier is queued because the threadpoolbulkhead has reached the maxThreadPoolSize the caller gets the nice timeout, but the supplier is still executed once a thread in the threadpoolbulkhead is available. I would have thought that any queued supplier would be cancelled and not ran. Hopefully this description makes sense.

Regards,
Scott
________________________________

**Comment by RobWin** [comment]
Yes, unfortunately the ExecutorService cannot cancel a task and remove it from the queue.
At least I'm not aware of it. :(
